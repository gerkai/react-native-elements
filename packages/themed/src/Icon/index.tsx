import { withTheme } from '../config';
import {
  Icon,
  IconProps,
  IconNode,
  IconObject,
  IconType,
} from '@react-native-elements/base/dist/Icon/Icon';

export { Icon };
export type { IconProps, IconNode, IconObject, IconType };
export default withTheme(Icon, 'Icon');
