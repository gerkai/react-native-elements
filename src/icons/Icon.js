import React, { PropTypes } from 'react'
import { Platform, TouchableHighlight, View, StyleSheet } from 'react-native'
import getIconType from '../helpers/getIconType'

let styles = {}

const Icon = ({
    type,
    name,
    size,
    color,
    iconStyle,
    component,
    onPress,
    underlayColor,
    reverse,
    raised,
    containerStyle,
    reverseColor
  }) => {
  let Component = View
  if (onPress) {
    Component = TouchableHighlight
  }
  if (component) {
    Component = component
  }
  let Icon
  if (!type) {
    Icon = getIconType('material')
  } else {
    Icon = getIconType(type)
  }
  return (
    <Component
      underlayColor={
        reverse ? color : underlayColor || color
      }
      style={[
        styles.button,
        raised && styles.raised, {
          borderRadius: size + 4,
          backgroundColor: reverse ? color : raised ? 'white' : 'transparent',
          height: size * 2 + 4,
          width: size * 2 + 4,
          alignItems: 'center',
          justifyContent: 'center'},
        containerStyle && containerStyle
      ]}
      onPress={onPress}>
      <Icon
        style={[
          {backgroundColor: 'transparent'},
          iconStyle && iconStyle
        ]}
        size={size}
        name={name}
        color={reverse ? reverseColor : color} />
    </Component>
  )
}

Icon.propTypes = {
  type: PropTypes.string,
  name: PropTypes.string,
  size: PropTypes.number,
  color: PropTypes.string,
  component: PropTypes.element,
  underlayColor: PropTypes.string,
  reverse: PropTypes.bool,
  raised: PropTypes.bool,
  containerStyle: PropTypes.object,
  iconStyle: PropTypes.object,
  onPress: PropTypes.func,
  reverseColor: PropTypes.string
}

Icon.defaultProps = {
  underlayColor: 'white',
  reverse: false,
  raised: false,
  size: 24,
  color: 'black',
  reverseColor: 'white'
}

styles = StyleSheet.create({
  button: {
    margin: 7
  },
  raised: {
    ...Platform.select({
      ios: {
        shadowColor: 'rgba(0,0,0, .4)',
        shadowOffset: {height: 1, width: 1},
        shadowOpacity: 1,
        shadowRadius: 1
      },
      android: {
        elevation: 2
      }
    })
  }
})

export default Icon
