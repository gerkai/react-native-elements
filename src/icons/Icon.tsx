import React from 'react';
import {
  Platform,
  TouchableHighlight,
  View,
  StyleSheet,
  TouchableNativeFeedback,
  ViewStyle,
  StyleProp,
  TextStyle,
  TouchableHighlightProps,
} from 'react-native';
import {
  IconButtonProps,
  IconProps as VectorIconProps,
} from 'react-native-vector-icons/Icon';
import Color from 'color';
import getIconType from '../helpers/getIconType';
import getIconStyle from '../helpers/getIconStyle';
import { withTheme } from '../config';
import { Theme } from '../config/theme';

export type IconType =
  | 'material'
  | 'material-community'
  | 'simple-line-icon'
  | 'zocial'
  | 'font-awesome'
  | 'octicon'
  | 'ionicon'
  | 'foundation'
  | 'evilicon'
  | 'entypo'
  | 'antdesign'
  | 'font-awesome-5'
  | string;

export interface IconObject extends TouchableHighlightProps {
  name?: string;
  color?: string;
  size?: number;
  type?: IconType;
  style?: StyleProp<TextStyle>;
}

export type IconNode = boolean | React.ReactElement<{}> | Partial<IconProps>;

export type IconProps = IconButtonProps & {
  type?: IconType;
  Component?: typeof React.Component;
  reverse?: boolean;
  raised?: boolean;
  containerStyle?: StyleProp<ViewStyle>;
  iconProps?: VectorIconProps;
  reverseColor?: string;
  disabled?: boolean;
  disabledStyle?: StyleProp<ViewStyle>;
  solid?: boolean;
  brand?: boolean;
  theme?: Theme;
};

const Icon: React.FunctionComponent<IconProps> = (props) => {
  const {
    type = 'material',
    name,
    size = 24,
    color: colorProp,
    iconStyle,
    iconProps,
    underlayColor = 'transparent',
    reverse = false,
    raised = false,
    containerStyle,
    reverseColor: reverseColorProp,
    disabled = false,
    disabledStyle,
    onPress,
    Component = onPress
      ? Platform.select<typeof React.Component>({
          android: TouchableNativeFeedback,
          default: TouchableHighlight,
        })
      : View,
    solid = false,
    brand = false,
    theme,
    ...attributes
  } = props;
  const color = colorProp || theme.colors.black;
  const reverseColor = reverseColorProp || theme.colors.white;
  const IconComponent = getIconType(type);
  const iconSpecificStyle = getIconStyle(type, { solid, brand });

  const getBackgroundColor = () => {
    if (reverse) {
      return color;
    }
    return raised ? theme.colors.white : 'transparent';
  };

  const buttonStyles = {
    borderRadius: size + 4,
    height: size * 2 + 4,
    width: size * 2 + 4,
  };

  if (Platform.OS === 'android' && !attributes.background) {
    if (Platform.Version >= 21) {
      // @ts-ignore
      attributes.background = TouchableNativeFeedback.Ripple(
        Color(color).alpha(0.2).rgb().string(),
        true
      );
    }
  }

  return (
    <View
      style={StyleSheet.flatten([
        styles.container,
        (reverse || raised) && styles.button,
        (reverse || raised) && buttonStyles,
        raised && styles.raised,
        iconStyle && iconStyle.borderRadius
          ? {
              borderRadius: iconStyle.borderRadius,
            }
          : {},
        containerStyle && containerStyle,
      ])}
    >
      <Component
        {...attributes}
        {...(onPress && {
          onPress,
          disabled,
          underlayColor: reverse ? color : underlayColor,
          activeOpacity: 0.3,
        })}
      >
        <View
          style={StyleSheet.flatten([
            (reverse || raised) && buttonStyles,
            {
              backgroundColor: getBackgroundColor(),
              alignItems: 'center',
              justifyContent: 'center',
            },
            disabled && styles.disabled,
            disabled && disabledStyle,
          ])}
        >
          <IconComponent
            testID="iconIcon"
            style={StyleSheet.flatten([
              { backgroundColor: 'transparent' },
              iconStyle && iconStyle,
            ])}
            size={size}
            name={name}
            color={reverse ? reverseColor : color}
            {...iconSpecificStyle}
            {...iconProps}
          />
        </View>
      </Component>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    overflow: 'hidden',
  },
  button: {
    margin: 7,
  },
  raised: {
    ...Platform.select({
      android: {
        elevation: 2,
      },
      default: {
        shadowColor: 'rgba(0,0,0, .4)',
        shadowOffset: { height: 1, width: 1 },
        shadowOpacity: 1,
        shadowRadius: 1,
      },
    }),
  },
  disabled: {
    backgroundColor: '#D1D5D8',
  },
});

export { Icon };
export default withTheme(Icon, 'Icon');
