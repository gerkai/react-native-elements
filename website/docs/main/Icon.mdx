---
id: icon

title: Icon

slug: /icon
---

import Usage from "./usage/Icon/Icon.mdx";

import { Icon } from "react-native-elements";

import Tabs from "@theme/Tabs";

import TabItem from "@theme/TabItem";

Icons are visual indicators usually used to describe action or intent.

They are also used for displaying information.

## Usage

<Usage />

## Props

### Icon

- [Component](#component)
- [backgroundColor](#backgroundcolor)
- [borderRadius](#borderradius)
- [brand](#brand)
- [color](#color)
- [containerStyle](#containerstyle)
- [disabled](#disabled)
- [disabledStyle](#disabledstyle)
- [iconProps](#iconprops)
- [iconStyle](#iconstyle)
- [name](#name)
- [raised](#raised)
- [reverse](#reverse)
- [reverseColor](#reversecolor)
- [size](#size)
- [solid](#solid)
- [style](#style)
- [type](#type)

## Reference

### Icon

#### Component

Update React Native Component.

| Type            | Default                                           |
| --------------- | ------------------------------------------------- |
| React Component | `Press handlers present then Pressable else View` |

---

#### backgroundColor

Background color of the button

| Type                 | Default   |
| -------------------- | --------- |
| number or ColorValue | '#007AFF' |

---

#### borderRadius

Border radius of the button

Set to 0 to disable.

| Type   | Default |
| ------ | ------- |
| number | 5       |

---

#### brand

Uses the brands font (FontAwesome5 only).

| Type    | Default |
| ------- | ------- |
| boolean | false   |

---

#### color

Text and icon color

Use iconStyle or nest a Text component if you need different colors.

| Type                 | Default |
| -------------------- | ------- |
| number or ColorValue | 'white' |

---

#### containerStyle

Add styling to container holding icon.

| Type               | Default |
| ------------------ | ------- |
| View style(Object) | None    |

---

#### disabled

If true, disable all interactions for this component.

Disables onPress events. Only works when `onPress` has a handler.

| Type    | Default |
| ------- | ------- |
| boolean | false   |

---

#### disabledStyle

Style for the button when disabled. Only works when `onPress` has a handler.

| Type               | Default |
| ------------------ | ------- |
| View style(Object) | None    |

---

#### iconProps

Provide all props from react-native Icon component.

| Type      | Default |
| --------- | ------- |
| IconProps | None    |

---

#### iconStyle

Styles applied to the icon only

Good for setting margins or a different color.

| Type      | Default           |
| --------- | ----------------- |
| TextStyle | {marginRight: 10} |

---

#### name

Name of the icon to show

See Icon Explorer app

{@link https ://github.com/oblador/react-native-vector-icons/tree/master/Examples/IconExplorer}

| Type   | Default |
| ------ | ------- |
| string | None    |

---

#### raised

Adds box shadow to button.

| Type    | Default |
| ------- | ------- |
| boolean | false   |

---

#### reverse

Reverses color scheme.

| Type    | Default |
| ------- | ------- |
| boolean | false   |

---

#### reverseColor

Specify reverse icon color.

| Type   | Default |
| ------ | ------- |
| string | None    |

---

#### size

Size of the icon, can also be passed as fontSize in the style object.

| Type   | Default |
| ------ | ------- |
| number | 12      |

---

#### solid

Uses the solid font.

| Type    | Default |
| ------- | ------- |
| boolean | false   |

---

#### style

Style prop inherited from TextProps and TouchableWithoutFeedbackProperties

Only exist here so we can have ViewStyle or TextStyle

| Type                   | Default |
| ---------------------- | ------- |
| ViewStyle or TextStyle | None    |

---

#### type

Type of icon set. [Supported sets here](#available-icon-sets).

| Type   | Default  |
| ------ | -------- |
| string | material |

---
