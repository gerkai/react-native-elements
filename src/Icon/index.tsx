import { withTheme } from '../config';
import { Icon, IconProps, IconNode, IconObject, IconType } from './Icon';

export { Icon };
export type { IconProps, IconNode, IconObject, IconType };
export default withTheme(Icon, 'Icon');
