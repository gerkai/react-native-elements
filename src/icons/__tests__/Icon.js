import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Icon from '../Icon';

describe('Icon component', () => {
  it('should render without issues', () => {
    const component = shallow(<Icon name="wifi" />);

    expect(component.length).toBe(1);
    expect(toJson(component)).toMatchSnapshot();
  });

  it('should render with icon type', () => {
    const component = shallow(
      <Icon
        name="alert"
        type="octicon"
        reverse
        color="red"
        iconStyle={{ backgroundColor: 'peru' }}
      />
    );

    expect(component.length).toBe(1);
    expect(toJson(component)).toMatchSnapshot();
  });

  it('should have onPress event', () => {
    const onPress = jest.fn();
    const component = shallow(<Icon onPress={onPress} name="wifi" />);
    const touchable = component.childAt(0);
    touchable.simulate('press');
    expect(onPress).toHaveBeenCalledTimes(1);
  });

  it('should apply default disabled styles', () => {
    const onPress = jest.fn();
    const component = shallow(<Icon onPress={onPress} name="wifi" disabled />);
    expect(toJson(component)).toMatchSnapshot();
  });

  it('should apply custom disabled styles', () => {
    const onPress = jest.fn();
    const component = shallow(
      <Icon
        onPress={onPress}
        name="wifi"
        disabled
        disabledStyle={{ backgroundColor: 'pink' }}
      />
    );
    expect(toJson(component)).toMatchSnapshot();
  });
});
