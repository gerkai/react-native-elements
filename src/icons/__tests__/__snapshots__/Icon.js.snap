// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Icon component should apply custom disabled styles 1`] = `
<Component>
  <TouchableHighlight
    activeOpacity={0.85}
    delayPressOut={100}
    disabled={true}
    onPress={[MockFunction]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "pink",
        "justifyContent": "center",
      }
    }
    underlayColor="white"
  >
    <Icon
      allowFontScaling={false}
      color="black"
      name="wifi"
      size={24}
      style={
        Object {
          "backgroundColor": "transparent",
        }
      }
    />
  </TouchableHighlight>
</Component>
`;

exports[`Icon component should apply default disabled styles 1`] = `
<Component>
  <TouchableHighlight
    activeOpacity={0.85}
    delayPressOut={100}
    disabled={true}
    onPress={[MockFunction]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#D1D5D8",
        "justifyContent": "center",
      }
    }
    underlayColor="white"
  >
    <Icon
      allowFontScaling={false}
      color="black"
      name="wifi"
      size={24}
      style={
        Object {
          "backgroundColor": "transparent",
        }
      }
    />
  </TouchableHighlight>
</Component>
`;

exports[`Icon component should render with icon type 1`] = `
<Component>
  <Component
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "red",
        "borderRadius": 28,
        "height": 52,
        "justifyContent": "center",
        "margin": 7,
        "width": 52,
      }
    }
    underlayColor="red"
  >
    <Icon
      allowFontScaling={false}
      color="white"
      name="alert"
      size={24}
      style={
        Object {
          "backgroundColor": "peru",
        }
      }
    />
  </Component>
</Component>
`;

exports[`Icon component should render without issues 1`] = `
<Component>
  <Component
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "transparent",
        "justifyContent": "center",
      }
    }
    underlayColor="white"
  >
    <Icon
      allowFontScaling={false}
      color="black"
      name="wifi"
      size={24}
      style={
        Object {
          "backgroundColor": "transparent",
        }
      }
    />
  </Component>
</Component>
`;
